### PR TITLE
Replace GivesExperience.FriendlyFire and GivesBounty.Stances with ValidStances

### LINKED ISSUE #11546: Replace GivesExperience.FriendlyFire boolean with Stances
`GivesBounty` already works that way.

Would make things more consistent and allow to disable giving experience for neutral actors, for example.

### PR DESCRIPTION
**Pull request by phrohdoh**
~~Depends on~~ #11482  has been merged.
Closes #11546

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/GivesExperience.cs [hunk 1]
```diff
@@ -19,8 +20,8 @@ class GivesExperienceInfo : ITraitInfo
 		[Desc("If -1, use the value of the unit cost.")]
 		public readonly int Experience = -1;
 
-		[Desc("Grant experience for team-kills.")]
-		public readonly bool FriendlyFire = false;
+		[Desc("Stance the attacking player needs to receive the experience.")]
+		public readonly Stance[] Stances = { Stance.Neutral, Stance.Enemy };
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/GivesExperience.cs]: Shouldn't `Stance ValidStances = Stance.Neutral | Stance.Enemy;` work as well (as we do elsewhere)?

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 2]
```diff
@@ -185,6 +185,23 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
 				}
 
+				// GivesExperience was changed from having a FriendlyFire bool to a Stances array.
+				if (engineVersion < 20160702 && node.Key.StartsWith("GivesExperience"))
+				{
+					var ff = "FriendlyFire";
+					var ffNode = node.Value.Nodes.FirstOrDefault(n => n.Key == ff);
+					if (ffNode != null)
+					{
+						if (FieldLoader.GetValue<bool>(ff, ffNode.Value.Value))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: It's quite unlikely to happen, but what if an actor has this set to `false` in order to overwrite an inherited `true`?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Then the node is removed and that is the end of the upgrade rule, leaving them with `Neutral` and `Enemy`.

File: OpenRA.Mods.Common/Traits/GivesExperience.cs (hunk 1 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/GivesExperience.cs]: Yes that also works, I just took this directly from `GivesBounty`.
I can change it but I'd like us to decide on a definitive way forward so this doesn't come up again.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 2 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Wouldn't the actor we inherit from (who had `true` before) get `Neutral, Enemy, Ally`?
And we inherit that then (although we only want `false`/`Neutral, Enemy`).

File: OpenRA.Mods.Common/Traits/GivesExperience.cs (hunk 1 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/GivesExperience.cs]: I'd vote for what I suggested.

**Comment by Phrohdoh** [comment]
Updated.

**Comment by reaperrr** [comment]
> OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:L209: [SA1108] A comment may not be placed within the bracketed statement.

**Comment by Phrohdoh** [comment]
Updated.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 3]
```diff
@@ -185,6 +185,33 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
+						var ffNode = node.Value.Nodes.FirstOrDefault(n => n.Key == ff);
+						if (ffNode != null)
+						{
+							var newStanceStr = "";
+							if (FieldLoader.GetValue<bool>(ff, ffNode.Value.Value))
+								newStanceStr = "Neutral, Enemy, Ally";
+							else
+								newStanceStr = "Neutral, Enemy";
+
+							node.Value.Nodes.Add(new MiniYamlNode("Stances", newStanceStr));
```
Comment (reaperrr) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: This needs to be "ValidStances" now.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Right-o, will fix.

**Comment by reaperrr** [comment]
Looks good to me :+1:

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

Remove the testcases now, please.

**Comment by Phrohdoh** [comment]
Ready for merge.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/ae7f86975193fb27fca3a5b21f6542f3b01f2c21...4445ba5352a97836eee06aad2d6355036914bf74)

**Pull request merged**